ocrd_models/ocrd_models/ocrd_mets.py

```diff
@@ -310,19 +310,36 @@ def add_file(self, fileGrp, mimetype=None, url=None, ID=None, pageId=None, force
             raise ValueError("Invalid syntax for mets:file/@ID %s (not an xs:ID)" % ID)
         if not REGEX_FILE_ID.fullmatch(fileGrp):
             raise ValueError("Invalid syntax for mets:fileGrp/@USE %s (not an xs:ID)" % fileGrp)
+        log = getLogger('ocrd_models.ocrd_mets.remove_file_group')
         el_fileGrp = self._tree.getroot().find(".//mets:fileGrp[@USE='%s']" % (fileGrp), NS)
         if el_fileGrp is None:
             el_fileGrp = self.add_file_group(fileGrp)
-        mets_file = next(self.find_files(ID=ID), None)
-        if mets_file and not ignore:
-            if not force:
-                raise Exception("File with ID='%s' already exists" % ID)
-            mets_file.url = url
-            mets_file.mimetype = mimetype
-            mets_file.ID = ID
-            mets_file.pageId = pageId
-            mets_file.local_filename = local_filename
-        else:
+        mets_file = None
+        if not ignore:
+            if pageId:
+                mets_file = next(self.find_files(fileGrp=fileGrp, mimetype=mimetype, pageId=pageId), None)
```

Comment:
I gave it some more thought. The insight was right, but I am not convinced anymore that it implies we must abolish the `add_file` resolution. Here's why:

In all cases except PAGE-XML files, we can have multiple files per page/grp/type. So there we'd always lay a trap: if we intervene, we might effectively create an artificial clash (e.g. by disallowing multiple textline images); if we don't, we might miss an actual clash (e.g. by allowing two competing ALTO-XML outputs).

So page/grp/type clashes cannot be decided.

Then how about ID clashes? In the benign case, the same kind of process that is responsible for the existing ID is now repeating. So we should be fine just checking `force/overwrite` mode (hence, reuse vs. raise). In the malignant case, some other process created the same ID, despite being unrelated by the matter (i.e. different page/grp/type). Suppose that we _can_ distinguish that case, how do we proceed? 

We cannot control the process' `make_file_id` (as we are just in a subsequent/subordinate call), and we cannot rename the existing ID either (as it might cause inconsistencies elsewhere). So I'd say we are forced to capitulate (i.e. raise).

That would amount to a much simplified decision tree. Am I missing something?

Reply:
> In all cases except PAGE-XML files, we can have multiple files per page/grp/type. So there we'd always lay a trap: if we intervene, we might effectively create an artificial clash (e.g. by disallowing multiple textline images); if we don't, we might miss an actual clash (e.g. by allowing two competing ALTO-XML outputs).

> So page/grp/type clashes cannot be decided.

91e5bd48a

First of all, I moved the ID clash check in add_file to the top. If ID clashes and neither `force` nor `ignore` are active, refuse to add outright.

I also got rid of the "reuse existing OcrdFile" mechanism altogether because it offers no benefit and can lead to hard-to-debug errors. Instead, in case of clashes and with `force` active, existing files will be removed first, before being added. No change in behavior for non-clashing files.

AFAICS there are two sets of files where straightforward clash detection won't work: a) AlternativeImages (which might be many with the same pageId, mimetype and filegrp) and b) document-wide documents like TEI or PDF that do not have a pageId.

To accomodate a), I've added a new kwarg `is_alternative_image` that, if set, will ignore the "grp/mimetype/pageId" check outright. Unless the IDs are clashing, it's perfectly valid to have many files with the same mimetype in the same fileGrp for the same pageId.

The files in b) have in common that they lack a pageId. So, in case there is a "grp/mimetype" clash without `pageId` being set, nothing is removed. This does prevent e.g. having two vartiants of TEI of the whole document in the same fileGrp. So if that is a realistic prospect, it needs another round of adaptions.

I **think** the decision tree in `OcrdMets.add_file` should largely reflect your proposal, with the added complication of handling `force`, `ignore`, `is_alternative_image` and "is a file without pageId" as described.

Reply:
> AFAICS there are two sets of files where straightforward clash detection won't work: a) AlternativeImages (which might be many with the same pageId, mimetype and filegrp) and b) document-wide documents like TEI or PDF that do not have a pageId.
> 
> To accomodate a), I've added a new kwarg `is_alternative_image` that, if set, will ignore the "grp/mimetype/pageId" check outright. Unless the IDs are clashing, it's perfectly valid to have many files with the same mimetype in the same fileGrp for the same pageId.

I see two problems with that:
1. you force the caller to decide in advance what the file is for. Hard to decide if you are in a shell script. (Basically, this extra arg has to be dragged through everywhere, with lots of places to adapt.)
2. there may be other multi-file-per-page cases besides derived images. Take custom formats like JSON or txt per region/line/word/glyph for example. We already have these cases in existing processors (e.g. ocrd-segment-extract-lines, ocrd-segment-replace-text).

Like I said, in all cases except PAGE-XML we simply do not know in advance whether multi-file semantics is ok or not. My formulation was more general and should not need that distinction. However, it hinges on the assumption that the malign case (ID clash from another, unrelated process) does not coincide with a grp/mime/page clash. (Which in my opinion is safe to make.)

> The files in b) have in common that they lack a pageId. So, in case there is a "grp/mimetype" clash without `pageId` being set, nothing is removed. This does prevent e.g. having two vartiants of TEI of the whole document in the same fileGrp. So if that is a realistic prospect, it needs another round of adaptions.

Yes, I agree we can assume that `pageId=None` is just as unique as any single pageId. But then you don't need the extra check. Notice this is already covered by my pseudocode (which makes for fewer conditionals).

Reply:
> I **think** the decision tree in `OcrdMets.add_file` should largely reflect your proposal, with the added complication of handling `force`, `ignore`, `is_alternative_image` and "is a file without pageId" as described.

If my analysis is right, then the only complication to my proposal would be the outer `if not ignore:`.

Reply:
Resolved, no need for (or rather, too much potential errors with) `is_alternative_image`, implemented largely as proposed.